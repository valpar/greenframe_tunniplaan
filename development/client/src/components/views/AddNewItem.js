import Modal from "../UI/Modal/Modal";
import NewSubject from "./inputRows/NewSubject";
import { useEffect, useState } from "react";
import axios from "axios";
import NewLecturer from "./inputRows/NewLecturer";
import NewRoom from "./inputRows/NewRoom";
import NewCourse from "./inputRows/NewCourse";
import ConfirmModal from "../UI/ConfirmModal/ConfirmModal";
import TooltipLarge from "../UI/Tooltip/TooltipLarge";
import content from "../../assets/content/content.json";
import RequestModal from "../UI/RequestModal/RequestModal";

axios.defaults.baseURL = `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`;

const AddNewItem = (props) => {
  const {
    modalFor,
    onNewItem,
    scheduled,
    roomsData,
    courseData,
    teacherData,
    subjectsData,
    onClose,
    onDelete,
    editMode,
    editValues,
  } = props;
  const {
    deleteMessage,
    roomHasActiveTeachers,
    courseHasActiveTeachers,
    teacherHasActiveTeachers,
    subjectHasActiveTeachers,
    saveMessage,
  } = content.confirmModalMessages;

  const token = JSON.parse(sessionStorage.getItem('token'));

  const { mandatoryFields } = content.errorMessages;

  const [inputsState, setInputsState] = useState([{}]);
  const [inputsAreValid, setInputsAreValid] = useState([{ inputs: false }]);
  const [validSubmit, setValidSubmit] = useState(true);
  const [responseId, setResponseId] = useState();
  const [showDeleteConfirmModal, setShowDeleteConfirmModal] = useState(false);
  const [showUpdateConfirmModal, setShowUpdateConfirmModal] = useState(false);
  const [deleteModalMessage, setDeleteModalMessage] = useState(deleteMessage);
  const [showOverlay, setOverlay] = useState(false);

  const [requestError, setRequestError] = useState(false);
  const [requestSuccess, setRequestSuccess] = useState(false);
  const [requestLoading, setRequestLoading] = useState(false);
  const [requestMessage, setRequestMessage] = useState("");
  const [showRequestModal, setShowRequestModal] = useState(false);
  const [hideModal, setHideModal] = useState(false);
  const [requestAction, setRequestAction] = useState("");

  const inputsChangeHandler = (inputsObj, rowIndex, validInputs) => {
    setInputsState((prevState) =>
      prevState.map((obj, i) => {
        if (i === rowIndex) return { ...inputsObj };
        return obj;
      })
    );
    setInputsAreValid((prevState) =>
      prevState.map((obj, i) => {
        if (i === rowIndex) return { inputs: validInputs };
        return obj;
      })
    );
  };
  useEffect(() => {
    if (!validSubmit)
      setValidSubmit(
        inputsAreValid.every((isValid) => isValid.inputs === true)
      );
  }, [inputsAreValid, validSubmit]);
  const addNewRowHandler = () => {
    setInputsState((prevState) => {
      return (prevState = [...prevState, {}]);
    });
    setInputsAreValid((prevState) => {
      return (prevState = [...prevState, {}]);
    });
  };
  const removeRowHandler = (index) => {
    setInputsState((prevState) => {
      return [...prevState.filter((e, i) => i !== index)];
    });
    setInputsAreValid((prevState) => {
      return [...prevState.filter((e, i) => i !== index)];
    });
  };

  const submitItemHandler = async () => {
    setOverlay(false);
    const isValid = inputsAreValid.every((isValid) => isValid.inputs === true);
    setValidSubmit(isValid);
    if (isValid) {
      let typeId;
      let responseId = [];
      if (modalFor === "rooms") typeId = "roomId";
      if (modalFor === "courses") typeId = "courseId";
      if (modalFor === "teachers") typeId = "teacherId";
      setRequestAction("create");
      try {
        setShowUpdateConfirmModal(false);
        setHideModal(true);
        setShowRequestModal(true);
        setRequestLoading(true);
        for (let state of inputsState) {
          await axios.post(`/${modalFor}`, { ...state }, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => {
            responseId.push({ [typeId]: response.data.id });
            if (modalFor === "subjects")
              onNewItem("subjectId", response.data.id);
            setResponseId(response.data.id);
          });
        }
      } catch (error) {
        setRequestLoading(false);
        setRequestError(true);
        setRequestMessage(content.errorMessages.requestAddError);
        return;
      }

      setRequestLoading(false);
      setRequestMessage(content.successMessages.create);
      setRequestSuccess(true);

      if (responseId.length > 0 && modalFor !== "subjects")
        onNewItem(modalFor, responseId);

      setInputsState([{}]);
      setInputsAreValid([{}]);
    }
  };

  const closeHandler = () => {
    if (modalFor === "subjects" && responseId) return onClose();
    if (responseId) return onClose();
    if (modalFor === "subjects") return onClose("subjectId");
    onClose(modalFor);
  };
  const confirmModalHandler = (event) => {
    if (event.target.name === "delete") {
      let roomBooked = scheduled.filter((row) => {
        if (row.rooms !== "") {
          let arr = row.rooms.filter(
            (room) => room.roomId === inputsState[0].id
          );
          return arr.length > 0 ? row : false;
        }
        return false;
      });
      let courseBooked = scheduled.filter((row) => {
        if (row.courses !== "") {
          let arr = row.courses.filter(
            (course) => course.courseId === inputsState[0].id
          );
          return arr.length > 0 ? row : false;
        }
        return false;
      });
      let teacherBooked = scheduled.filter((row) => {
        if (row.teachers !== "") {
          let arr = row.teachers.filter(
            (teacher) => teacher.teacherId === inputsState[0].id
          );
          return arr.length > 0 ? row : false;
        }
        return false;
      });
      let subjectBooked = scheduled.filter((row) => {
        if (row.subject !== "") {
          return row.subject.id === inputsState[0].id ? row : false;
        }
        return false;
      });

      if (modalFor === "rooms" && roomBooked.length > 0)
        setDeleteModalMessage(roomHasActiveTeachers);
      if (modalFor === "courses" && courseBooked.length > 0)
        setDeleteModalMessage(courseHasActiveTeachers);
      if (modalFor === "teachers" && teacherBooked.length > 0)
        setDeleteModalMessage(teacherHasActiveTeachers);
      if (modalFor === "subject" && subjectBooked.length > 0)
        setDeleteModalMessage(subjectHasActiveTeachers);

      setShowDeleteConfirmModal(true);
      setOverlay(true);
    }

    if (event.target.name !== "delete") {
      const isValid = inputsAreValid.every(
        (isValid) => isValid.inputs === true
      );
      if (!isValid) return setValidSubmit(isValid);
      if (event.target.name === "update") setShowUpdateConfirmModal(true);
      if (event.target.name === "create") setShowUpdateConfirmModal(true);
      setOverlay(true);
    }
  };

  const declineHandler = () => {
    setOverlay(false);
    setShowDeleteConfirmModal(false);
  };

  const declineUpdateHandler = () => {
    setOverlay(false);
    setShowUpdateConfirmModal(false);
  };

  const updateItemHandler = async () => {
    setOverlay(false);
    setRequestAction("update");
    try {
      setShowUpdateConfirmModal(false);
      setOverlay(false);
      setHideModal(true);
      setShowRequestModal(true);
      setRequestLoading(true);
      if (modalFor === "rooms") {
        await axios
          .patch(`/rooms/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "courses") {
        await axios
          .patch(`/courses/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "teachers") {
        await axios
          .patch(`/teachers/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "subjects") {
        await axios
          .patch(`/subjects/${editValues}`, inputsState[0], {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
    } catch (error) {
      setRequestLoading(false);
      setRequestError(true);
      setRequestMessage(content.errorMessages.requestUpdateError);
      return;
    }

    setRequestLoading(false);
    setRequestMessage(content.successMessages.update);
    setRequestSuccess(true);
  };

  const deleteItemHandler = async () => {
    setRequestAction("delete");
    try {
      setShowDeleteConfirmModal(false);
      setOverlay(false);
      setHideModal(true);
      setShowRequestModal(true);
      setRequestLoading(true);
      if (modalFor === "rooms") {
        await axios
          .delete(`/rooms/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "courses") {
        await axios
          .delete(`/courses/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "teachers") {
        await axios
          .delete(`/teachers/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
      if (modalFor === "subjects") {
        await axios
          .delete(`/subjects/${editValues}`, {
            headers: { Authorization: `Bearer ${token.token}` },
          })
          .catch((error) =>  {
            if (error.response.status === 401) {
              setShowRequestModal(false);
              props.forceLogoutHandler();
              return error;
            }
            return error;
          })
          .then((response) => console.log(response));
      }
    } catch (error) {
      setRequestLoading(false);
      setRequestError(true);
      setRequestMessage(content.errorMessages.requestDeleteError);
      return;
    }
    setRequestLoading(false);
    setRequestMessage(content.successMessages.delete);
    setRequestSuccess(true);
    setDeleteModalMessage("KUSTUTA");
  };
  useEffect(() => {
    if (!validSubmit) {
      const timer = setTimeout(() => {
        setValidSubmit(true);
      }, 1000);
      return () => clearTimeout(timer);
    }
  }, [validSubmit]);

  useEffect(() => {
    if (requestSuccess) {
      const timer = setTimeout(() => {
        if (
          requestMessage === content.successMessages.delete ||
          requestMessage === content.successMessages.update
        ) {
          onDelete();
        }
        onClose();
        setHideModal(false);
        setShowRequestModal(false);
        setRequestSuccess(false);
        setRequestMessage("");
      }, 2000);
      return () => clearTimeout(timer);
    }
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [requestSuccess]);

  const endRequestHandler = () => {
    onClose();
  };
  const failedRequestConfirmHandler = () => {
    setRequestError(false);
    setRequestMessage("");
    if (requestAction === "create") submitItemHandler();
    if (requestAction === "delete") deleteItemHandler();
    if (requestAction === "update") updateItemHandler();
  };

  return (
    <Modal onHidden={hideModal} onClose={closeHandler} overlay={showOverlay}>
      <div className="relative flex flex-col lg:pl-4">
        <div className="relative flex justify-end">
          <i
            onClick={closeHandler}
            className={`bi bi-x-lg absolute text-3xl lg:text-xl -top-2 -right-2`}
          ></i>
        </div>
        {modalFor === "subjects" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewSubject
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  subjectsData={subjectsData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}
        {modalFor === "teachers" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewLecturer
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  teacherData={teacherData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}
        {modalFor === "courses" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewCourse
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  courseData={courseData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}
        {modalFor === "rooms" &&
          inputsState.map((inputsRow, i) => {
            return (
              <div key={i}>
                <NewRoom
                  editValues={editValues}
                  editMode={editMode}
                  onAddNewRow={addNewRowHandler}
                  onRemoveRow={removeRowHandler}
                  modalFor={modalFor}
                  onChange={inputsChangeHandler}
                  index={i}
                  roomsData={roomsData}
                  values={inputsState[i]}
                  count={inputsState.length}
                />
              </div>
            );
          })}

        <div
          className={`flex ${
            editMode
              ? "justify-between space-x-20"
              : "justify-center lg:justify-between"
          } w-full pt-8`}
        >
          <div className="relative">
            {showDeleteConfirmModal && (
              <div className="absolute top-20 -left-16">
                <ConfirmModal
                  onDecline={declineHandler}
                  onConfirm={deleteItemHandler}
                  modalMessage={deleteModalMessage}
                  topArrow={true}
                />
              </div>
            )}
            {editMode && (
              <button
                onClick={confirmModalHandler}
                className="btn-actions"
                type="button"
                name="delete"
              >
                KUSTUTA
              </button>
            )}
          </div>
          <div className="relative">
            {!validSubmit && (
              <div className="absolute -top-16 -left-10 w-[150%]">
                <TooltipLarge message={mandatoryFields} />
              </div>
            )}

            <button
              onClick={confirmModalHandler}
              className="btn-actions"
              type="submit"
              name={editMode ? "update" : "create"}
            >
              SALVESTA
            </button>

            {validSubmit && showUpdateConfirmModal && (
              <div className="absolute top-20 -left-16">
                <ConfirmModal
                  onDecline={declineUpdateHandler}
                  onConfirm={editMode ? updateItemHandler : submitItemHandler}
                  modalMessage={saveMessage}
                  topArrow={true}
                />
              </div>
            )}
          </div>
        </div>
        {showRequestModal && (
          <RequestModal
            error={requestError}
            success={requestSuccess}
            loading={requestLoading}
            modalMessage={requestMessage}
            customStyle="top-1/2 lg:ml-32"
            onDecline={endRequestHandler}
            onConfirm={failedRequestConfirmHandler}
          />
        )}
      </div>
    </Modal>
  );
};

export default AddNewItem;
