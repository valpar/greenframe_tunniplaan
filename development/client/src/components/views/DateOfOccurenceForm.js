import {
  formatHoursMinutes,
  formatMilliseconds,
} from "../../utils/Format/Date";
import InputWithLabel from "../UI/Input/InputWithLabel";
import { useEffect, useState } from "react";
import DropdownInput from "../UI/Dropdown/DropdownInput";
import CalendarOneInput from "../UI/Calendar/CalendarOneInput";
import content from "../../assets/content/content.json";

const loadCalculator = (load, startTime) => {
  const minutes = [45, 90, 150, 195, 300, 345, 405, 450, 510, 555, 615, 660];
  const splittedStartTime = startTime.split(":");
  const minutesTotal =
    splittedStartTime[0] * 60 +
    parseInt(splittedStartTime[1]) +
    minutes[load - 1];
  const pad = (input, hour) => {
    if (hour && input === 24) return "00";
    if (hour && input > 24)
      return input - 24 < 10 ? "0" + (input - 24) : input - 24;
    return input < 10 ? "0" + input : input;
  };
  return `${pad(Math.floor(minutesTotal / 60), "hour")}:${pad(
    Math.floor(minutesTotal % 60)
  )}`;
};

const beginValues = [
  {
    label: "10:00",
    value: "10:00",
  },
  {
    label: "14:15",
    value: "14:15",
  },
  {
    label: "17:30",
    value: "17:30",
  },
  {
    label: "18:00",
    value: "18:00",
  },
];
const endValues = [
  {
    label: "13:15",
    value: "13:15",
  },
  {
    label: "17:30",
    value: "17:30",
  },
  {
    label: "18:00",
    value: "18:00",
  },
  {
    label: "20:00",
    value: "20:00",
  },
];

const lunchValues = [
  {
    label: "JAH",
    value: "JAH",
  },
  {
    label: "EI",
    value: "EI",
  },
];

const DateOfOccurenceForm = (props) => {
  const [dateValue, setDateValue] = useState();
  const [showCalendar, setShowCalendar] = useState(false);
  const [loadValue, setLoadValue] = useState("");
  const [startTime, setStartTime] = useState("10:00");
  const [endTime, setEndTime] = useState("");
  const [lunchValue, setLunchValue] = useState("");
  const [calendarReset, setCalendarReset] = useState(false);
  const [enteredDateIsValid, setEnteredDateIsValid] = useState(false);
  const [enteredLoadIsValid, setEnteredLoadIsValid] = useState(false);
  const [enteredStartTimeIsValid, setEnteredStartTimeIsValid] = useState(false);
  const [enteredEndTimeIsValid, setEnteredEndTimeIsValid] = useState(false);
  const [errorMessages, setErrorMessages] = useState({
    date: "",
    load: "",
    startTime: "",
    endTime: "",
  });
  const { isMobile, isTabletOrMobile } = props;

  const { requiresDate, requiresEndDate, notUniqueDate } =
    content.errorMessages;

  useEffect(() => {
    if (props.onNotValidFields[props.index]) {
      setEnteredEndTimeIsValid(!props.onNotValidFields[props.index].endTime);
      if (!props.onNotValidFields[props.index].endTime) {
        setErrorMessages((prevState) => {
          return {
            ...prevState,
            endTime: requiresEndDate,
          };
        });
      }

      if (!props.onNotValidFields[props.index].date) {
        setEnteredDateIsValid(!props.onNotValidFields[props.index].date);
        setErrorMessages((prevState) => {
          return {
            ...prevState,
            date: notUniqueDate,
          };
        });
      }
      if (!dateValue) {
        setEnteredDateIsValid(true);
        setErrorMessages((prevState) => {
          return {
            ...prevState,
            date: requiresEndDate,
          };
        });
      }
    }
    if (props.onNotValidFields[props.index]?.date && dateValue) {
      setEnteredDateIsValid(false);
      setErrorMessages((prevState) => {
        return {
          ...prevState,
          date: "",
        };
      });
    }
  }, [props.onNotValidFields, props.index]);

  useEffect(() => {
    if (props.onAfterSubmit) {
      setLoadValue("");
      setStartTime("10:00");
      setEndTime("");
      setDateValue(undefined);
      setLunchValue("");
      setCalendarReset(true);
    }
  }, [props.onAfterSubmit]);

  const addDateHandler = (event) => {
    event.preventDefault();
    setShowCalendar((prevState) => (prevState = !prevState));
  };
  const loadChangeHandler = (load) => {
    setLoadValue(load);
    setEnteredLoadIsValid(false);
    if (!load.match(/(^0?[1-9]$)|(^1[0-2]$)/)) {
      setEnteredLoadIsValid(true);
    }
  };
  const calendarClickHandler = (date) => {
    if (!date) {
      setDateValue(undefined);
      return;
    }
    setCalendarReset(false);
    setEnteredDateIsValid(false);
    setErrorMessages((prevState) => {
      return {
        ...prevState,
        date: "",
      };
    });
    if (!(formatMilliseconds(date) > formatMilliseconds(new Date()))) {
      setEnteredDateIsValid(true);
    }

    setDateValue(date);
    if (startTime.length > 0) {
      let valueArr = startTime.split(":");
      props.onChange(
        [
          {
            startTime: `${new Date(
              date.setUTCHours(valueArr[0], valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
    }
    if (endTime.length > 0) {
      let valueArr = endTime.split(":");
      props.onChange(
        [
          {
            endTime: `${new Date(
              date.setUTCHours(valueArr[0], valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
    }
    setShowCalendar(false);
  };

  const startTimeChangeHandler = (value) => {
    setCalendarReset(false);
    setEnteredStartTimeIsValid(false);
    if (value.match(/^(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$/)) {
      let valueArr = value.split(":");
      setStartTime(value);
      props.onChange(
        [
          {
            startTime: `${new Date(
              dateValue.setUTCHours(valueArr[0], valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
    } else {
      setEnteredStartTimeIsValid(true);
      setStartTime(value);
      props.onChange(
        [
          {
            startTime: "",
          },
        ],
        props.index
      );
    }
  };
  const endTimeChangeHandler = (value) => {
    setEnteredEndTimeIsValid(false);
    if (value !== "") {
      setErrorMessages((prevState) => {
        return {
          ...prevState,
          endTime: "",
        };
      });
    }

    if (value.match(/^(0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$/)) {
      let valueArr = value.split(":");
      setEndTime(value);
      props.onChange(
        [
          {
            endTime: `${new Date(
              dateValue.setUTCHours(valueArr[0], valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
    } else {
      setEnteredEndTimeIsValid(true);
      setEndTime(value);
      props.onChange(
        [
          {
            endTime: "",
          },
        ],
        props.index
      );
    }
  };

  const lunchChangeHandler = () => {
    let valueArr = endTime.split(":");
    if (endTime > "14:59" && lunchValue === "JAH") {
      setLunchValue("EI");

      props.onChange(
        [
          {
            endTime: `${new Date(
              dateValue.setUTCHours(valueArr[0] - 1, valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
      setEndTime((prevState) => {
        let valueArr = prevState.split(":");
        return `${valueArr[0] - 1}:${valueArr[1]}`;
      });
    }
    if (endTime > "13:59" && lunchValue === "EI") {
      setLunchValue("JAH");
      props.onChange(
        [
          {
            endTime: `${new Date(
              dateValue.setUTCHours(+valueArr[0] + 1, valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
      setEndTime((prevState) => {
        let valueArr = prevState.split(":");
        return `${+valueArr[0] + 1}:${valueArr[1]}`;
      });
    }
  };

  useEffect(() => {
    if (dateValue) {
      let valueArr = startTime.split(":");
      props.onChange(
        [
          {
            startTime: `${new Date(
              dateValue.setUTCHours(valueArr[0], valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
    }
  }, []);

  useEffect(() => {
    if (loadValue > 0 && loadValue < 13 && startTime.length === 5) {
      setEndTime(loadCalculator(loadValue, startTime));
      let valueArr = loadCalculator(loadValue, startTime).split(":");
      props.onChange(
        [
          {
            endTime: `${new Date(
              dateValue.setUTCHours(valueArr[0], valueArr[1], "00", "000")
            ).toISOString()}`,
          },
        ],
        props.index
      );
    }
    if (loadValue > 4) setLunchValue("JAH");
    if (loadValue < 5) setLunchValue("EI");
    if ((loadValue === "" || loadValue > 12) && endTime.length === 0) {
      setEndTime("");
      props.onChange(
        [
          {
            endTime: "",
          },
        ],
        props.index
      );
    }
  }, [loadValue, startTime]);

  const deleteRowHandler = () => {
    props.onDelete(props.index);
  };
  useEffect(() => {
    if (props.editMode) {
      setDateValue(new Date(props.editData.startTime));
      setStartTime(formatHoursMinutes(props.editData.startTime));
      setEndTime(formatHoursMinutes(props.editData.endTime));
    }
  }, []);

  return (
    <div data-testid='date-of-occurence-form' className="flex flex-col w-full items-center lg:items-end lg:flex-row lg:space-x-4">
      <div className="flex flex-col lg:flex-row w-full space-y-3 lg:space-x-4 lg:space-y-0">
        <div className="lg:w-full lg:min-w-[15rem]">
          <CalendarOneInput
            onShowCalendar={showCalendar}
            onClickDay={calendarClickHandler}
            value={dateValue}
            index={props.index}
            onClick={addDateHandler}
            label="KuupÃ¤ev"
            hasError={enteredDateIsValid}
            errorMessage={errorMessages.date}
            reset={calendarReset}
            isMobile={isMobile}
            isDesktop={isTabletOrMobile}
          />
        </div>
        <div className="lg:w-full">
          <InputWithLabel
            data-testid={`input-${props.name}`} // et ei oleks mitu sama nimega data-testid
            onChange={loadChangeHandler}
            type="text"
            name="workLoad"
            label="Maht"
            value={loadValue}
            index={props.index}
            hasError={enteredLoadIsValid}
            readOnly={!dateValue ? true : false}
            onErrorMessage={!dateValue ? requiresDate : errorMessages.load}
            isMobile={isMobile}
            tooltipWidth="w-full lg:w-52 z-20"
            isDesktop={isTabletOrMobile}
          />
        </div>

        <div className="lg:w-full">
          <DropdownInput
            onLoad={startTimeChangeHandler}
            onChange={startTimeChangeHandler}
            options={beginValues}
            value={startTime}
            label="Algus"
            name="startTime"
            cssClass="dropdownOccurence"
            index={props.index}
            hasError={enteredStartTimeIsValid}
            readOnly={!dateValue ? true : false}
            showOptions={dateValue ? true : false}
            onErrorMessage={!dateValue ? requiresDate : errorMessages.startTime}
            isMobile={isMobile}
            isTabletOrMobile={isTabletOrMobile}
            isDesktop={isTabletOrMobile}
          />
        </div>

        <div className="lg:w-full">
          <DropdownInput
            onChange={endTimeChangeHandler}
            options={endValues}
            label="LÃµpp"
            cssClass="dropdownOccurence"
            name="endTime"
            value={endTime}
            index={props.index}
            hasError={enteredEndTimeIsValid}
            readOnly={!dateValue ? true : false}
            showOptions={dateValue ? true : false}
            onErrorMessage={!dateValue ? requiresDate : errorMessages.endTime}
            isTabletOrMobile={isTabletOrMobile}
            isDesktop={isTabletOrMobile}
          />
        </div>

        <div className="lg:w-full">
          <InputWithLabel
            readOnly={true}
            onClick={lunchChangeHandler}
            label="LÃµuna"
            cssClass="dropdownOccurence"
            name="hasLunch"
            value={lunchValue}
            index={props.index}
            onErrorMessage={!endTime ? requiresEndDate : ""}
            isMobile={isMobile}
            tooltipWidth="w-full lg:w-52 z-20"
            customStyle="lg:text-center cursor-pointer"
            isDesktop={isTabletOrMobile}
          />
        </div>
      </div>

      {!props.editMode && (
        <div className="flex justify-center items-center space-x-10 pt-4 lg:pt-0  lg:pb-1 lg:justify-end !lg:items-end w-18 lg:space-x-4">
          {props.index === props.occurenceLength - 1 && !props.editMode && (
            <i
              onClick={props.onClick}
              className={`bi bi-plus-lg text-4xl lg:text-2xl text-right`}
            ></i>
          )}
          {props.occurenceLength > 1 && (
            <i
              onClick={deleteRowHandler}
              className={` bi bi-x-lg text-3xl lg:text-xl text-center`}
            ></i>
          )}
        </div>
      )}
    </div>
  );
};

export default DateOfOccurenceForm;
