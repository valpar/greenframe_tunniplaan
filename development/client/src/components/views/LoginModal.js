import React, { Fragment, useEffect } from "react";
import ReactDOM from "react-dom";
import Input from "../UI/Input/Input";

const Backdrop = (props) => {
  useEffect(() => {
    const themeColorMeta = document.querySelector('meta[name="theme-color"]');

    themeColorMeta.setAttribute("content", "#5C5C5C");

    return () => {
      themeColorMeta.setAttribute("content", "#fff");
    };
  }, []);

  return (
    <div
      data-testid="backdrop"
      className="fixed top-0 left-0 w-full h-screen z-20 bg-neutral-900 opacity-70"
      onClick={props.onClose}
    />
  );
};

const ConfirmModalOverlay = (props) => {
  
  const changeHandler = (input) => {
    
    if (input.name==="email") {
      props.onEmailChange(input.value);
    }
    if (input.name==="password") {
      props.onPasswordChange(input.value);
    }
  }
  

  let arrow =
    "absolute z-40 w-8 h-8 -rotate-45 -right-4 bg-white top-1/2 -translate-y-1/2";
  if (props.topArrow)
    arrow =
      "absolute z-40 w-8 h-8 -rotate-45 -top-4 bg-white left-1/2 -translate-x-1/2";
  if (props.bottomArrow)
    arrow =
      "absolute z-40 w-8 h-8 -rotate-45 -bottom-4 bg-white left-1/2 -translate-x-1/2";

  return (
    <div className="fixed top-1/2 -translate-y-1/2 left-1/2 -translate-x-1/2 lg:translate-x-0 lg:translate-y-0 lg:left-auto lg:top-auto lg:relative w-auto h-auto space-y-2 bg-white p-4 rounded z-30 shadow-lg">
      <div className={`hidden lg:block ${arrow}`}></div>
      <div className="pb-4 text-center">
        <p>{props.modalMessage}</p>
      </div>

      <label>
        E-mail: 
        <Input
          name="email" 
          value={props.email} 
          onChange={changeHandler} 
          readOnly={false}
          className={`p-2 border ${
            props.hasError || props.errorMessage
              ? "border-red-500"
              : "border-borderBlack "
          } rounded-none font-normal focus:outline-none focus:border-black`}
        />
      </label>
      <label>
        Parool: 
        <Input 
          name="password" 
          value={props.password} 
          onChange={changeHandler} 
          type="password" readOnly={false} 
          className={`p-2 border ${
            props.hasError || props.errorMessage
              ? "border-red-500"
              : "border-borderGray "
          } rounded-none font-normal focus:outline-none focus:border-black`}
        />
      </label>


      <div className="flex space-x-4">
        <button
          onClick={props.onDecline}
          type="button"
          className="border border-borderGray py-1 px-2 w-24 hover:bg-borderGray"
        >
          {props.modalButtons ? props.modalButtons[0] : `EI`}
        </button>
        <button
          onClick={props.onConfirm}
          type="button"
          className="border border-borderGray py-1 px-2 w-24 hover:bg-borderGray"
        >
          {props.modalButtons ? props.modalButtons[1] : `JAH`}
        </button>
      </div>
    </div>
  );
};



const portalElement = document.getElementById("overlays");

const LoginModal = (props) => {

  return (
    
    <Fragment>
      {ReactDOM.createPortal(<Backdrop />, portalElement)}
      <ConfirmModalOverlay {...props} /> 
    </Fragment>
    
  );
};

export default LoginModal;
