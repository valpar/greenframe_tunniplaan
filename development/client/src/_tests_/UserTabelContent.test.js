import React from 'react';
import { render, fireEvent, waitFor, screen, cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { UsersTabelContent } from '../components/views/UsersTabelContent';


describe('UserTabelContent', () => {
  // peab mockima userlisti, sest muidu ei loo snapshotti, kuna return users.map Cannot read properties of undefined 
  it('matches snapshot', () => {
    const mockUsers = [
      { firstName: 'John', lastName: 'Doe', email: 'john@example.com', role: 'admin' }
    ];
    const mockRoleOptions = [{ label: 'Admin', value: 'admin' }];
    const { tree } = renderer.create(
      <UsersTabelContent 
        users={mockUsers} 
        roleOptions={mockRoleOptions}
      />
    ).toJSON;
    expect(tree).toMatchSnapshot();
  });
});
