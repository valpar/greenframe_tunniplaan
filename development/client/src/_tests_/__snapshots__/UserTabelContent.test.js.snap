// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTabelContent matches snapshot 1`] = `undefined`;
